presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java

```diff
@@ -106,153 +131,171 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
                 {
-                    return new LdapPrincipal(user);
+                    return principal;
                 }
             }, servletResponse);
         }
         catch (AuthenticationException e) {
-            // log the failure since the exception goes directly to the client
+            log.debug(e, "LDAP authentication failed");
+            if (e.getStatus() == UNAUTHORIZED) {
+                sendChallenge(request, response, e.getMessage());
+                return;
+            }
+            response.sendError(e.getStatus().code(), e.getMessage());
+        }
+    }
+
+    private static void sendChallenge(HttpServletRequest request, HttpServletResponse response, String errorMessage)
+            throws IOException
+    {
+        long contentLength = request.getContentLengthLong();
+        if (contentLength > 0) {
+            skipFully(request.getInputStream(), contentLength);
+        }
+        response.setHeader(HttpHeaders.WWW_AUTHENTICATE, "Basic realm=\"presto\"");
+        response.sendError(UNAUTHORIZED.code(), errorMessage);
+    }
+
+    private static Credentials getCredentials(String header)
+            throws AuthenticationException
+    {
+        if (header == null) {
+            throw new AuthenticationException(UNAUTHORIZED, "Unauthorized");
+        }
+        if (!header.startsWith(BASIC_AUTHENTICATION_PREFIX)) {
+            throw new AuthenticationException(BAD_REQUEST, "Basic authentication is expected");
+        }
+        String base64Credentials = header.substring(BASIC_AUTHENTICATION_PREFIX.length());
+        String credentials = new String(Base64.getDecoder().decode(base64Credentials), UTF_8);
+        List<String> parts = Splitter.on(':').limit(2).splitToList(credentials);
+        if (parts.size() != 2 || parts.stream().anyMatch(String::isEmpty)) {
+            throw new AuthenticationException(BAD_REQUEST, "Invalid credentials: " + credentials);
+        }
+        return new Credentials(parts.get(0), parts.get(1));
+    }
+
+    private Principal getPrincipal(Credentials credentials)
+            throws AuthenticationException
+    {
+        try {
+            return authenticationCache.get(credentials);
+        }
+        catch (ExecutionException e) {
+            Throwable cause = e.getCause();
+            propagateIfInstanceOf(cause, AuthenticationException.class);
+            throw Throwables.propagate(cause);
+        }
+    }
+
+    private Principal authenticate(Credentials credentials)
+            throws AuthenticationException
+    {
+        Map<String, String> environment = createEnvironment(credentials);
+        String user = credentials.getUser();
+        InitialDirContext context = null;
+        try {
+            context = createDirContext(environment);
+            checkForGroupMembership(user, context);
+            log.debug("Authentication successful for user %s.", user);
+            return new LdapPrincipal(user);
```

Comment:
I just think that such interface is better. Like `authenticate` method accepts `Credentials` and returns `Principal` if the authentication is successful. I believe it is more clear than just return `String`.